starport/services/scaffolder/type.go

```diff
@@ -13,45 +12,97 @@ import (
 	"github.com/tendermint/starport/starport/pkg/xgenny"
 	modulecreate "github.com/tendermint/starport/starport/templates/module/create"
 	"github.com/tendermint/starport/starport/templates/typed"
+	"github.com/tendermint/starport/starport/templates/typed/basic"
 	"github.com/tendermint/starport/starport/templates/typed/indexed"
 	"github.com/tendermint/starport/starport/templates/typed/singleton"
 )
 
-type TypeModel string
+// AddTypeOption configures options for AddType.
+type AddTypeOption func(*addTypeOptions)
 
-const (
-	List      TypeModel = "list"
-	Map       TypeModel = "map"
-	Singleton TypeModel = "singleton"
-)
+type addTypeOptions struct {
+	moduleName string
+	fields     []string
+
+	isList      bool
```

Comment:
With the TypeModel way, we don't have a way of having a default. So, this change is about making it optional to specify a kind for AddType. The default, only scaffolds the type without anything extra.

Reply:
Mmh we can have a default with a construction for the struct or by using a iota

Reply:
Mmh we can have a default by defining a constructor or by using a iota enum

Reply:
I couldn't get what you mean by the constructor part but with iota option, we still need to pass a value to the AddType. By default, I meant that you don't need to pass anything to AddType to get the default behaviour.

Reply:
For constructor I mean the method always used to initialize the type, currently we have:
```
o := addTypeOptions{
	moduleName: path.Package, // app's default module.
}
```
The type cannot be created without specifying a module so we have some value to determine before initializing the option anyway. For me we should have a constructor with module name and type model as required parameter.
Booleans seem confusing because they make feel like list or map are additional characteristic and you can combine them

Reply:
I cannot see why module name should be required, because we chose to use the default one if no name is specified. Still cannot get what you mean by the constructor option, maybe I would better undestand if you share the signature on your mind for the AddType func?

I agree that we cannot allow to use for ex. `List()` option with `Map()` together, and using it would be error prone. In this case the last passed option will overwrite previous ones in the current impl. Also, using them together would be a programmer mistake and in such case we should panic when you pass `Map()` in this example.

In general, functional options always has this drawback because some options may not be meant to be used to together. Same can happen to CLI flags to give an example for a similar situation. So, we should always check and panic.

If you still don't quite like this option and want to pass it as a seperate argument to AddTypes, instead of iota way, we can make user to pass one of the `List()`, `Map(withIndexes)`, `Single()` and `Type`. Because this way, for the Map kind, AddType's signature can provide a clean API to pass custom indexes. ​


So,  in this case the API would become from:

```go
func (s *Scaffolder) AddType(typeName string, tracer *placeholder.Tracer, options ...AddTypeOption) 
```

to:
```go
func (s *Scaffolder) AddType(typeName string, tracer *placeholder.Tracer, kind AddTypeKind, options ...AddTypeOption) 
```

And usage would stay the same, only you must pass type kind before options.

```go
AddType("user", tracer, Map(email, age), TypeWithModule("non-default-module")
```

Reply:
I applied the propositon in my previous message, please let me know if that works for you. 08dcb41

Reply:
This boolean is not set or used by the callers of AddType func. Are you suggesting to change the public API of AddType or the internal implementation?

Reply:
only in the internal implementation, to be more clear whats is the object type

Reply:
I don't see how boolean creates a disadvantage, we use it in the switch stmt which is what would you also do if you used iota. If you feel this is important please open a counter PR with the improvement on your mind and we can go from there.

Reply:
The disadvantage is that the code is harder to understand because using booleans would imply the type can be a list or a map at the same time while those are disjointed characteristic. So by using an enum you prevent, by design, the type from being two different things at the same time.

This is honestly a small concern in a small chunk of code, I don't think it is worth to spend too much time discussing about this, I just give my opinion